File: server_py/app/services/draconic_ancestry.py
from __future__ import annotations

from typing import Any
from math import floor


DRACONIC_ANCESTRIES: tuple[dict[str, str], ...] = (
    {"id": "black", "label": "Black", "damageType": "acid", "resistanceType": "acid", "breathWeaponShape": "line", "breathWeaponSaveType": "dexterity", "breathWeaponAreaSize": "1.5m x 9m"},
    {"id": "blue", "label": "Blue", "damageType": "lightning", "resistanceType": "lightning", "breathWeaponShape": "line", "breathWeaponSaveType": "dexterity", "breathWeaponAreaSize": "1.5m x 9m"},
    {"id": "brass", "label": "Brass", "damageType": "fire", "resistanceType": "fire", "breathWeaponShape": "line", "breathWeaponSaveType": "dexterity", "breathWeaponAreaSize": "1.5m x 9m"},
    {"id": "bronze", "label": "Bronze", "damageType": "lightning", "resistanceType": "lightning", "breathWeaponShape": "line", "breathWeaponSaveType": "dexterity", "breathWeaponAreaSize": "1.5m x 9m"},
    {"id": "copper", "label": "Copper", "damageType": "acid", "resistanceType": "acid", "breathWeaponShape": "line", "breathWeaponSaveType": "dexterity", "breathWeaponAreaSize": "1.5m x 9m"},
    {"id": "gold", "label": "Gold", "damageType": "fire", "resistanceType": "fire", "breathWeaponShape": "cone", "breathWeaponSaveType": "constitution", "breathWeaponAreaSize": "4.5m"},
    {"id": "green", "label": "Green", "damageType": "poison", "resistanceType": "poison", "breathWeaponShape": "cone", "breathWeaponSaveType": "constitution", "breathWeaponAreaSize": "4.5m"},
    {"id": "red", "label": "Red", "damageType": "fire", "resistanceType": "fire", "breathWeaponShape": "cone", "breathWeaponSaveType": "constitution", "breathWeaponAreaSize": "4.5m"},
    {"id": "silver", "label": "Silver", "damageType": "cold", "resistanceType": "cold", "breathWeaponShape": "cone", "breathWeaponSaveType": "constitution", "breathWeaponAreaSize": "4.5m"},
    {"id": "white", "label": "White", "damageType": "cold", "resistanceType": "cold", "breathWeaponShape": "cone", "breathWeaponSaveType": "constitution", "breathWeaponAreaSize": "4.5m"},
)

DRACONIC_ANCESTRY_BY_ID: dict[str, dict[str, str]] = {
    ancestry["id"]: ancestry for ancestry in DRACONIC_ANCESTRIES
}

DRACONIC_ANCESTRY_DAMAGE_TYPES: dict[str, str] = {
    ancestry["id"]: ancestry["damageType"] for ancestry in DRACONIC_ANCESTRIES
}

DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY = "draconicAncestry"
LEGACY_DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY = "dragonAncestor"


def is_valid_draconic_ancestry(value: object) -> bool:
    return isinstance(value, str) and value in DRACONIC_ANCESTRY_DAMAGE_TYPES


def get_draconic_damage_type(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    return DRACONIC_ANCESTRY_DAMAGE_TYPES.get(ancestry)


def get_draconic_ancestry_label(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    entry = DRACONIC_ANCESTRY_BY_ID.get(ancestry)
    if entry is None:
        return None
    return entry["label"]


def get_draconic_resistance_type(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    entry = DRACONIC_ANCESTRY_BY_ID.get(ancestry)
    if entry is None:
        return None
    return entry["resistanceType"]


def get_draconic_breath_weapon_shape(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    entry = DRACONIC_ANCESTRY_BY_ID.get(ancestry)
    if entry is None:
        return None
    return entry["breathWeaponShape"]


def get_draconic_breath_weapon_save_type(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    entry = DRACONIC_ANCESTRY_BY_ID.get(ancestry)
    if entry is None:
        return None
    return entry["breathWeaponSaveType"]


def get_draconic_breath_weapon_area_size(ancestry: str | None) -> str | None:
    if ancestry is None:
        return None
    entry = DRACONIC_ANCESTRY_BY_ID.get(ancestry)
    if entry is None:
        return None
    return entry["breathWeaponAreaSize"]


def get_draconic_ancestry_from_data(data: object) -> str | None:
    if not isinstance(data, dict):
        return None
    normalized_config = normalize_subclass_config(
        data.get("subclass"),
        data.get("subclassConfig"),
    )
    ancestry = (
        normalized_config.get(DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY)
        if normalized_config is not None
        else None
    )
    return ancestry if is_valid_draconic_ancestry(ancestry) else None


def resolve_draconic_lineage_state(data: object) -> dict[str, Any]:
    if not isinstance(data, dict):
        return {
            "ancestry": None,
            "damageType": None,
            "resistanceType": None,
            "hasElementalAffinity": False,
            "resistances": [],
        }

    if data.get("class") != "sorcerer" or data.get("subclass") != "draconic_bloodline":
        return {
            "ancestry": None,
            "damageType": None,
            "resistanceType": None,
            "hasElementalAffinity": False,
            "resistances": [],
        }

    ancestry = get_draconic_ancestry_from_data(data)
    damage_type = get_draconic_damage_type(ancestry)
    resistance_type = get_draconic_resistance_type(ancestry)
    level = int(data.get("level", 1) or 1)
    has_elemental_affinity = level >= 6 and damage_type is not None and resistance_type is not None

    return {
        "ancestry": ancestry,
        "ancestryLabel": get_draconic_ancestry_label(ancestry),
        "damageType": damage_type,
        "resistanceType": resistance_type,
        "hasElementalAffinity": has_elemental_affinity,
        "resistances": [resistance_type] if has_elemental_affinity and resistance_type else [],
    }


def resolve_elemental_affinity(data: object, spell_damage_type: object) -> dict[str, Any]:
    lineage = resolve_draconic_lineage_state(data)
    normalized_damage_type = (
        str(spell_damage_type).strip().lower()
        if isinstance(spell_damage_type, str) and spell_damage_type.strip()
        else None
    )
    charisma_score = 10
    if isinstance(data, dict):
        abilities = data.get("abilities")
        if isinstance(abilities, dict) and isinstance(abilities.get("charisma"), int):
            charisma_score = abilities["charisma"]
    eligible = bool(
        lineage["hasElementalAffinity"]
        and lineage["damageType"]
        and normalized_damage_type
        and lineage["damageType"] == normalized_damage_type
    )
    return {
        "eligible": eligible,
        "damageType": lineage["damageType"],
        "bonus": floor((charisma_score - 10) / 2) if eligible else None,
    }


def normalize_subclass_config(
    subclass: object,
    subclass_config: object,
) -> dict[str, str] | None:
    if not isinstance(subclass_config, dict):
        return None

    next_config: dict[str, str] = {}
    for raw_key, raw_value in subclass_config.items():
        if not isinstance(raw_key, str) or not isinstance(raw_value, str):
            continue
        value = raw_value.strip()
        if not value:
            continue
        key = (
            DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY
            if raw_key == LEGACY_DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY
            else raw_key
        )
        next_config[key] = value

    if subclass != "draconic_bloodline":
        next_config.pop(DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY, None)

    return next_config or None


def validate_draconic_subclass_state(data: object) -> tuple[bool, str | None]:
    if not isinstance(data, dict):
        return True, None

    if data.get("subclass") != "draconic_bloodline":
        return True, None

    normalized_config = normalize_subclass_config(
        data.get("subclass"),
        data.get("subclassConfig"),
    )
    ancestry = (
        normalized_config.get(DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY)
        if normalized_config is not None
        else None
    )

    if ancestry is None:
        return False, "draconicAncestry is required for Draconic Bloodline"

    if not is_valid_draconic_ancestry(ancestry):
        return False, "draconicAncestry is invalid for Draconic Bloodline"

    return True, None

File: apps/control-web/src/features/character-sheet/data/draconicAncestry.ts
export const DRACONIC_ANCESTRIES = [
  { id: "black", label: "Negro", damageLabel: "Ácido", damageType: "acid", resistanceType: "acid", breathWeaponShape: "line", breathWeaponSaveType: "dexterity", breathWeaponAreaSize: "1.5m x 9m" },
  { id: "blue", label: "Azul", damageLabel: "Elétrico", damageType: "lightning", resistanceType: "lightning", breathWeaponShape: "line", breathWeaponSaveType: "dexterity", breathWeaponAreaSize: "1.5m x 9m" },
  { id: "brass", label: "Latão", damageLabel: "Fogo", damageType: "fire", resistanceType: "fire", breathWeaponShape: "line", breathWeaponSaveType: "dexterity", breathWeaponAreaSize: "1.5m x 9m" },
  { id: "bronze", label: "Bronze", damageLabel: "Elétrico", damageType: "lightning", resistanceType: "lightning", breathWeaponShape: "line", breathWeaponSaveType: "dexterity", breathWeaponAreaSize: "1.5m x 9m" },
  { id: "copper", label: "Cobre", damageLabel: "Ácido", damageType: "acid", resistanceType: "acid", breathWeaponShape: "line", breathWeaponSaveType: "dexterity", breathWeaponAreaSize: "1.5m x 9m" },
  { id: "gold", label: "Ouro", damageLabel: "Fogo", damageType: "fire", resistanceType: "fire", breathWeaponShape: "cone", breathWeaponSaveType: "constitution", breathWeaponAreaSize: "4.5m" },
  { id: "green", label: "Verde", damageLabel: "Veneno", damageType: "poison", resistanceType: "poison", breathWeaponShape: "cone", breathWeaponSaveType: "constitution", breathWeaponAreaSize: "4.5m" },
  { id: "red", label: "Vermelho", damageLabel: "Fogo", damageType: "fire", resistanceType: "fire", breathWeaponShape: "cone", breathWeaponSaveType: "constitution", breathWeaponAreaSize: "4.5m" },
  { id: "silver", label: "Prata", damageLabel: "Frio", damageType: "cold", resistanceType: "cold", breathWeaponShape: "cone", breathWeaponSaveType: "constitution", breathWeaponAreaSize: "4.5m" },
  { id: "white", label: "Branco", damageLabel: "Frio", damageType: "cold", resistanceType: "cold", breathWeaponShape: "cone", breathWeaponSaveType: "constitution", breathWeaponAreaSize: "4.5m" },
] as const;

export type DraconicAncestry = (typeof DRACONIC_ANCESTRIES)[number];
export type DraconicAncestryId = DraconicAncestry["id"];
export type DraconicDamageType = DraconicAncestry["damageType"];
export type DraconicResistanceType = DraconicAncestry["resistanceType"];
export type DraconicBreathWeaponShape = DraconicAncestry["breathWeaponShape"];
export type DraconicBreathWeaponSaveType = DraconicAncestry["breathWeaponSaveType"];
export type DraconicLineageState = {
  ancestry: DraconicAncestryId | null;
  ancestryLabel: string | null;
  damageType: DraconicDamageType | null;
  resistanceType: DraconicResistanceType | null;
  hasElementalAffinity: boolean;
  resistances: DraconicResistanceType[];
};

type DraconicLineageInput = {
  classId?: string | null | undefined;
  subclass?: string | null | undefined;
  level?: number | null | undefined;
  subclassConfig?: Record<string, string> | null | undefined;
};

const DRACONIC_ANCESTRY_BY_ID = Object.fromEntries(
  DRACONIC_ANCESTRIES.map((ancestry) => [ancestry.id, ancestry]),
) as Record<DraconicAncestryId, DraconicAncestry>;

export const DRACONIC_ANCESTRY_DAMAGE_TYPES = Object.fromEntries(
  DRACONIC_ANCESTRIES.map((ancestry) => [ancestry.id, ancestry.damageType]),
) as Record<DraconicAncestryId, DraconicDamageType>;

export const DRACONIC_ANCESTRY_OPTIONS: Array<{
  id: DraconicAncestryId;
  name: string;
}> = DRACONIC_ANCESTRIES.map((ancestry) => ({
  id: ancestry.id,
  name: `${ancestry.label} (${ancestry.damageLabel})`,
}));

export const DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY = "draconicAncestry";
export const LEGACY_DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY = "dragonAncestor";

export const isValidDraconicAncestry = (
  value: string | null | undefined,
): value is DraconicAncestryId =>
  !!value && value in DRACONIC_ANCESTRY_DAMAGE_TYPES;

export const getDraconicAncestry = (
  ancestry: string | null | undefined,
): DraconicAncestry | null =>
  isValidDraconicAncestry(ancestry) ? DRACONIC_ANCESTRY_BY_ID[ancestry] : null;

export const getDraconicAncestryLabel = (
  ancestry: string | null | undefined,
): string | null =>
  getDraconicAncestry(ancestry)?.label ?? null;

export const getDraconicDamageType = (
  ancestry: string | null | undefined,
): DraconicDamageType | null =>
  getDraconicAncestry(ancestry)?.damageType ?? null;

export const getDraconicResistanceType = (
  ancestry: string | null | undefined,
): DraconicResistanceType | null =>
  getDraconicAncestry(ancestry)?.resistanceType ?? null;

export const getDraconicBreathWeaponShape = (
  ancestry: string | null | undefined,
): DraconicBreathWeaponShape | null =>
  getDraconicAncestry(ancestry)?.breathWeaponShape ?? null;

export const getDraconicBreathWeaponSaveType = (
  ancestry: string | null | undefined,
): DraconicBreathWeaponSaveType | null =>
  getDraconicAncestry(ancestry)?.breathWeaponSaveType ?? null;

export const getDraconicBreathWeaponAreaSize = (
  ancestry: string | null | undefined,
): string | null =>
  getDraconicAncestry(ancestry)?.breathWeaponAreaSize ?? null;

export const getDraconicLineageState = ({
  classId,
  subclass,
  level,
  subclassConfig,
}: DraconicLineageInput): DraconicLineageState => {
  const normalizedSubclass = String(subclass ?? "").trim().toLowerCase();
  const normalizedClassId = String(classId ?? "").trim().toLowerCase();
  if (normalizedClassId && normalizedClassId !== "sorcerer") {
    return {
      ancestry: null,
      ancestryLabel: null,
      damageType: null,
      resistanceType: null,
      hasElementalAffinity: false,
      resistances: [],
    };
  }
  if (normalizedSubclass !== "draconic_bloodline") {
    return {
      ancestry: null,
      ancestryLabel: null,
      damageType: null,
      resistanceType: null,
      hasElementalAffinity: false,
      resistances: [],
    };
  }

  const ancestryValue = subclassConfig?.[DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY];
  const ancestry = isValidDraconicAncestry(ancestryValue) ? ancestryValue : null;
  const ancestryData = ancestry ? getDraconicAncestry(ancestry) : null;
  const damageType = ancestryData?.damageType ?? null;
  const resistanceType = ancestryData?.resistanceType ?? null;
  const hasElementalAffinity = Number(level ?? 0) >= 6 && !!damageType;

  return {
    ancestry,
    ancestryLabel: ancestryData?.label ?? null,
    damageType,
    resistanceType,
    hasElementalAffinity,
    resistances: hasElementalAffinity && resistanceType ? [resistanceType] : [],
  };
};

export const resolveElementalAffinityEligibility = ({
  classId,
  subclass,
  level,
  subclassConfig,
  spellDamageType,
  charismaScore,
}: DraconicLineageInput & {
  spellDamageType?: string | null | undefined;
  charismaScore?: number | null | undefined;
}) => {
  const lineage = getDraconicLineageState({ classId, subclass, level, subclassConfig });
  const normalizedSpellDamageType = String(spellDamageType ?? "").trim().toLowerCase() || null;
  const eligible = Boolean(
    lineage.hasElementalAffinity &&
    lineage.damageType &&
    normalizedSpellDamageType &&
    lineage.damageType === normalizedSpellDamageType,
  );

  return {
    eligible,
    damageType: eligible ? lineage.damageType : lineage.damageType,
    bonus:
      eligible && typeof charismaScore === "number"
        ? Math.floor((charismaScore - 10) / 2)
        : null,
  };
};

export const normalizeSubclassConfig = (
  subclass: string | null | undefined,
  subclassConfig: Record<string, string> | null | undefined,
): Record<string, string> | null => {
  if (!subclassConfig) {
    return null;
  }

  const nextConfig = Object.fromEntries(
    Object.entries(subclassConfig)
      .map(([key, value]) => [
        key === LEGACY_DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY
          ? DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY
          : key,
        value,
      ])
      .filter(([, value]) => typeof value === "string" && value.trim().length > 0),
  ) as Record<string, string>;

  if (subclass !== "draconic_bloodline") {
    delete nextConfig[DRACONIC_ANCESTRY_SUBCLASS_CONFIG_KEY];
  }

  return Object.keys(nextConfig).length > 0 ? nextConfig : null;
};
